PayloadPkg/OsLoader/OsLoader.c

```diff
@@ -628,13 +575,303 @@ StartBooting (
+  OUT EFI_HANDLE          *FsHandle
+  )
+{
+  EFI_STATUS      Status;
+  UINT32          SwPart;
+  UINT32          FsType;
+  INT32           BootSlot;
+
+  ASSERT (OsBootOption != NULL);
+
+  SwPart = OsBootOption->SwPart;
+  FsType = OsBootOption->FsType;
+
+  DEBUG ((DEBUG_INFO, "Init File system\n"));
+  if ((OS_FILE_SYSTEM_TYPE)FsType < EnumFileSystemMax) {
+    Status = InitFileSystem (SwPart, FsType, HwPartHandle, FsHandle);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_INFO, "No partitions found, Status = %r\n", Status));
+      return Status;
+    }
+  } else {
+    //
+    // Assume RAW format partition
+    //
+    DEBUG ((DEBUG_INFO, "Requested FsType %d. Assume RAW format partition.\n", FsType));
+    *FsHandle = NULL;
+  }
+
+  //
+  // Get boot image A/B info. TBD: Better to be in LoadBootImages ()
+  //
+  BootSlot = GetBootSlot (OsBootOption, HwPartHandle);
+  DEBUG ((DEBUG_INFO, "BootSlot = 0x%x\n", BootSlot));
+  if (BootSlot == 1) {
+    OsBootOption->BootFlags |= LOAD_IMAGE_FROM_BACKUP;
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Parse a Boot Image
+
+  Based on given loaded image, this function will parse the image
+
+  @param[in]      OsBootOption  OS boot option to boot
+  @param[in, out] LoadedImage   Loaded Boot Image to be parsed
+
+  @retval  EFI_SUCCESS          No issue when parsing the image
+  @retval  Others               An error when parsing the image
+
+**/
+EFI_STATUS
+EFIAPI
+ParseBootImages (
+  IN  OS_BOOT_OPTION    *OsBootOption,
+  IN  EFI_HANDLE         LoadedImageHandle
+  )
+{
+  LOADED_IMAGE        *LoadedImage;
+  EFI_STATUS           Status;
+  UINT8                Type;
+
+  Status = EFI_SUCCESS;
+  for (Type = 0; Type < LoadImageTypeMax; Type++) {
+    if (Type == LoadImageTypeMisc) {
+      continue;
+    }
+
+    GetLoadedImageByType (LoadedImageHandle, Type, &LoadedImage);
+    if (LoadedImage == NULL) {
+      continue;
+    }
+
+    DEBUG ((DEBUG_INFO, "ParseBootImage ImageType-%d\n", Type));
+    if ((LoadedImage->Flags & LOADED_IMAGE_CONTAINER) != 0) {
+      if (FeaturePcdGet (PcdContainerBootEnabled)) {
+        Status = ParseContainerImage (OsBootOption, LoadedImage);
+      }
+    } else if ((LoadedImage->Flags & LOADED_IMAGE_IAS) != 0) {
+      Status = ParseIasImage (OsBootOption, LoadedImage);
+    }
+
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_INFO, "ParseLoadedImage: Status = %r\n", Status));
+      break;
+    }
+  }
+
+  return Status;
+}
+
+EFI_STATUS
+EFIAPI
+SetupBootImages (
+  IN  OS_BOOT_OPTION    *OsBootOption,
+  IN  EFI_HANDLE         LoadedImageHandle
+  )
+{
+  LOADED_IMAGE        *LoadedImage;
+  LOADED_IMAGE        *LoadedTrustyImage;
+  EFI_STATUS           Status;
+
+  GetLoadedImageByType (LoadedImageHandle, LoadImageTypeNormal, &LoadedImage);
+  GetLoadedImageByType (LoadedImageHandle, LoadImageTypeTrusty, &LoadedTrustyImage);
+
+  if (LoadedImage == NULL && LoadedTrustyImage == NULL) {
+    return EFI_LOAD_ERROR;
+  }
```

Comment:
Normal OS image is always required, and trusty image is optional. so just check LoadedImage should be fine if we want to check.

Reply:
Right, let me update this.

Reply:
Updated as commented.